# Objectives
    EXPLAIN, check performance of query
    solving by:
        add index
        union all

In [1]:
%load_ext sql
%sql mysql+pymysql://root:d9982951@localhost/employees

## query performace

In [2]:
%%sql
    USE employees

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.


[]

In [3]:
%%sql
    SET profiling = 1;
    SELECT *
    FROM employees;
    SHOW PROFILES;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
300024 rows affected.
3 rows affected.


[(1, 7.8e-05, 'commit'),
 (2, 5.159601, 'SELECT *\n    FROM employees'),
 (3, 6.7e-05, 'commit')]

300024 rows in 0.72sec

In [4]:
%%sql
    EXPLAIN
    SELECT *
    FROM employees;

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[(1, 'SIMPLE', 'employees', None, 'ALL', None, None, None, None, 299866, 100.0, None)]

### clean profile

In [76]:
%%sql
    SET @@profiling = 0;
    SET @@profiling_history_size = 0;
    SET @@profiling_history_size = 100; 
    SET @@profiling = 1;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
0 rows affected.
0 rows affected.
0 rows affected.


[]

## Use add index method

    show index

In [5]:
%%sql
    SHOW INDEX FROM employees;

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[('employees', 0, 'PRIMARY', 1, 'emp_no', 'A', 299866, None, None, '', 'BTREE', '', '')]

    query condition

In [6]:
%%sql
    SET profiling = 1;
    SELECT *
    FROM employees
    WHERE hire_date >= '2000-01-01';
    SHOW PROFILES;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
13 rows affected.
12 rows affected.


[(1, 7.8e-05, 'commit'),
 (2, 5.159601, 'SELECT *\n    FROM employees'),
 (3, 6.7e-05, 'commit'),
 (4, 6.3e-05, 'commit'),
 (5, 0.000291, 'EXPLAIN\n    SELECT *\n    FROM employees'),
 (6, 0.000113, 'commit'),
 (7, 0.005023, 'SHOW INDEX FROM employees'),
 (8, 8.1e-05, 'commit'),
 (9, 0.000127, 'SET profiling = 1'),
 (10, 0.000105, 'commit'),
 (11, 0.126501, "SELECT *\n    FROM employees\n    WHERE hire_date >= '2000-01-01'"),
 (12, 0.000956, 'commit')]

In [26]:
%%sql
    EXPLAIN
    SELECT *
    FROM employees
    WHERE hire_date >= '2000-01-01';

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[(1, 'SIMPLE', 'employees', None, 'ALL', None, None, None, None, 299866, 33.33, 'Using where')]

In [27]:
%%sql
    select *
    from employees
    limit 3;

 * mysql+pymysql://root:***@localhost/employees
3 rows affected.


[(10001, datetime.date(1953, 9, 2), 'Georgi', 'Facello', 'M', datetime.date(1986, 6, 26)),
 (10002, datetime.date(1964, 6, 2), 'Bezalel', 'Simmel', 'F', datetime.date(1985, 11, 21)),
 (10003, datetime.date(1959, 12, 3), 'Parto', 'Bamford', 'M', datetime.date(1986, 8, 28))]

    create index into hire date

In [28]:
%%sql
    CREATE INDEX hire_date_index ON employees(hire_date);

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.


[]

In [29]:
%%sql
    SHOW INDEX FROM employees;

 * mysql+pymysql://root:***@localhost/employees
2 rows affected.


[('employees', 0, 'PRIMARY', 1, 'emp_no', 'A', 299866, None, None, '', 'BTREE', '', ''),
 ('employees', 1, 'hire_date_index', 1, 'hire_date', 'A', 5139, None, None, '', 'BTREE', '', '')]

In [32]:
%%sql
    SET profiling=1;
    SELECT *
    FROM employees
    WHERE hire_date >= '2000-01-01';
    SHOW PROFILES;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
13 rows affected.
6 rows affected.


[(60, 0.000428, 'commit'),
 (61, 0.000231, 'commit'),
 (62, 0.000213, 'SET profiling=1'),
 (63, 8.8e-05, 'commit'),
 (64, 0.000603, "SELECT *\n    FROM employees\n    WHERE hire_date >= '2000-01-01'"),
 (65, 0.000212, 'commit')]

    DROP INDEX

In [33]:
%%sql
    DROP INDEX hire_date_index ON employees;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.


[]

In [35]:
%%sql
    set profiling=1;
    select *
    from employees
    where hire_date >= '2000-01-01';
    show profiles;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
13 rows affected.
6 rows affected.


[(70, 4.4e-05, 'commit'),
 (71, 4.6e-05, 'commit'),
 (72, 0.000132, 'set profiling=1'),
 (73, 7.3e-05, 'commit'),
 (74, 0.101483, "select *\n    from employees\n    where hire_date >= '2000-01-01'"),
 (75, 6.4e-05, 'commit')]

    the index improve performance of query in:
        0.101483 > 0.000603

## use UNION ALL method

In [63]:
%%sql
    SET profiling=1;
    SELECT *
    FROM employees
    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%';
    SHOW PROFILES;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
28970 rows affected.
5 rows affected.


[(166, 5.8e-05, 'commit'),
 (167, 0.000207, 'SET profiling=1'),
 (168, 8.1e-05, 'commit'),
 (169, 0.211455, "SELECT *\n    FROM employees\n    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%'"),
 (170, 7.6e-05, 'commit')]

In [64]:
%%sql
    EXPLAIN
    SELECT *
    FROM employees
    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%';

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[(1, 'SIMPLE', 'employees', None, 'ALL', None, None, None, None, 299866, 20.99, 'Using where')]

    add index to first name and last name

In [65]:
%%sql
    CREATE INDEX first_name_index ON employees(first_name);
    CREATE INDEX last_name_index ON employees(last_name);

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
0 rows affected.


[]

In [69]:
%%sql
    SET profiling=1;
    SELECT *
    FROM employees
    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%';
    SHOW PROFILES;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
28970 rows affected.
15 rows affected.


[(180, 0.000165, 'commit'),
 (181, 0.000245, 'SET profiling=1'),
 (182, 8e-05, 'commit'),
 (183, 0.234022, "SELECT *\n    FROM employees\n    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%'"),
 (184, 7.7e-05, 'commit'),
 (185, 0.000101, 'commit'),
 (186, 0.000132, 'SET profiling=1'),
 (187, 9.3e-05, 'commit'),
 (188, 0.141226, "SELECT *\n    FROM employees\n    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%'"),
 (189, 7.6e-05, 'commit'),
 (190, 7.3e-05, 'commit'),
 (191, 0.000115, 'SET profiling=1'),
 (192, 0.000147, 'commit'),
 (193, 0.137151, "SELECT *\n    FROM employees\n    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%'"),
 (194, 6.5e-05, 'commit')]

In [70]:
%%sql
    EXPLAIN
    SELECT *
    FROM employees
    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%';

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[(1, 'SIMPLE', 'employees', None, 'ALL', 'first_name_index,last_name_index', None, None, None, 299866, 20.99, 'Using where')]

    using union all
        see explain row scanning
        see running time

In [74]:
%%sql
    set profiling=1;
    
    SELECT * FROM employees WHERE first_name LIKE 'C%' UNION ALL SELECT * FROM employees WHERE last_name LIKE 'C%';
    
    show profiles;

 * mysql+pymysql://root:***@localhost/employees
0 rows affected.
29730 rows affected.
15 rows affected.


[(198, 0.000108, 'commit'),
 (199, 0.000147, 'set profiling=1'),
 (200, 9.1e-05, 'commit'),
 (201, 0.17242, "SELECT * FROM employees WHERE first_name LIKE 'C%' UNION ALL SELECT * FROM employees WHERE last_name LIKE 'C%'"),
 (202, 6.5e-05, 'commit'),
 (203, 0.000103, 'commit'),
 (204, 0.000145, 'set profiling=1'),
 (205, 9.2e-05, 'commit'),
 (206, 0.159324, "SELECT * FROM employees WHERE first_name LIKE 'C%' UNION ALL SELECT * FROM employees WHERE last_name LIKE 'C%'"),
 (207, 7.6e-05, 'commit'),
 (208, 5.9e-05, 'commit'),
 (209, 0.000161, 'set profiling=1'),
 (210, 0.000153, 'commit'),
 (211, 0.139555, "SELECT * FROM employees WHERE first_name LIKE 'C%' UNION ALL SELECT * FROM employees WHERE last_name LIKE 'C%'"),
 (212, 6.5e-05, 'commit')]

In [75]:
%%sql
    explain
    SELECT * FROM employees WHERE first_name LIKE 'C%' UNION ALL SELECT * FROM employees WHERE last_name LIKE 'C%';

 * mysql+pymysql://root:***@localhost/employees
2 rows affected.


[(1, 'PRIMARY', 'employees', None, 'range', 'first_name_index', 'first_name_index', '16', None, 20622, 100.0, 'Using index condition'),
 (2, 'UNION', 'employees', None, 'range', 'last_name_index', 'last_name_index', '18', None, 34168, 100.0, 'Using index condition')]

    after union all
    normal select change

In [77]:
%%sql
    EXPLAIN
    SELECT *
    FROM employees
    WHERE first_name LIKE 'C%' OR last_name LIKE 'C%';

 * mysql+pymysql://root:***@localhost/employees
1 rows affected.


[(1, 'SIMPLE', 'employees', None, 'ALL', 'first_name_index,last_name_index', None, None, None, 299866, 20.99, 'Using where')]